core/processor/ProcessorMergeMultilineLogNative.cpp

```diff
@@ -0,0 +1,445 @@
+                    // EndPatternReg不存在，StartPattern不存在，ContinuePatternReg没匹配上(理论上不支持这种正则配置)
+                    // 把multiBeginIndex到curIndex-1的日志合并到multiBeginIndex中
+                    mergeEvents.pop_back();
+                    MergeEvents(mergeEvents, true);
+                    sourceEvents[newEventsSize++] = std::move(sourceEvents[multiBeginIndex]);
+                    multiBeginIndex = curIndex;
+                    HandleUnmatchLogs(sourceEvents, multiBeginIndex, curIndex, newEventsSize, logPath);
+                    mergeEvents.clear();
+                    state = SPLIT_UNMATCH;
+                }
+                break;
+        }
+        if (!exception.empty()) {
+            if (AppConfig::GetInstance()->IsLogParseAlarmValid()) {
+                const StringView& logPath = logGroup.GetMetadata(EventGroupMetaKey::LOG_FILE_PATH_RESOLVED);
+                if (GetContext().GetAlarm().IsLowLevelAlarmValid()) {
+                    LOG_ERROR(GetContext().GetLogger(),
+                              ("regex_match in LogSplit fail, exception", exception)("project",
+                                                                                     GetContext().GetProjectName())(
+                                  "logstore", GetContext().GetLogstoreName())("file", logPath));
+                }
+                GetContext().GetAlarm().SendAlarm(REGEX_MATCH_ALARM,
+                                                  "regex_match in LogSplit fail:" + exception + ", file"
+                                                      + logPath.to_string(),
+                                                  GetContext().GetProjectName(),
+                                                  GetContext().GetLogstoreName(),
+                                                  GetContext().GetRegion());
+            }
+        }
+    }
+    // We should clear the log from `multiBeginIndex` to `size`.
+    if (multiBeginIndex < sourceEvents.size()) {
+        if (mMultiline.GetStartPatternReg() != NULL && mMultiline.GetEndPatternReg() == NULL) {
+            // If logs is unmatched, they have been handled immediately. So logs must be matched here.
+            MergeEvents(mergeEvents, true);
+            sourceEvents[newEventsSize++] = std::move(sourceEvents[multiBeginIndex]);
+        } else if (mMultiline.GetStartPatternReg() == NULL && mMultiline.GetContinuePatternReg() == NULL
+                   && mMultiline.GetEndPatternReg() != NULL) {
+            // If there is still logs in cache, it means that there is no end line. We can handle them as unmatched.
+            HandleUnmatchLogs(sourceEvents, multiBeginIndex, sourceEvents.size() - 1, newEventsSize, logPath, true);
+            mergeEvents.clear();
+        } else {
+            HandleUnmatchLogs(sourceEvents, multiBeginIndex, sourceEvents.size() - 1, newEventsSize, logPath, true);
+            mergeEvents.clear();
+        }
+    }
+    sourceEvents.resize(newEventsSize);
+}
```

Comment:
我重构了一下，可以看下测试用例是否通过和齐全
```suggestion
void ProcessorMergeMultilineLogNative::MergeLogsByRegex(PipelineEventGroup& logGroup) {
    auto& sourceEvents = logGroup.MutableEvents();
    size_t begin = 0, newSize = 0;
    std::vector<LogEvent*> events;

    std::string exception;
    bool isPartialLog = false;
    if (mMultiline.GetStartPatternReg() == nullptr && mMultiline.GetContinuePatternReg() == nullptr) {
        // if only end pattern is given, then it will stick to this state
        isPartialLog = true;
    }
    for (size_t cur = 0; cur < sourceEvents.size(); ++cur) {
        if (!IsSupportedEvent(sourceEvents[cur])) {
            for (size_t i = begin; i < sourceEvents.size(); ++i) {
                sourceEvents[newSize++] = std::move(sourceEvents[i]);
            }
            sourceEvents.resize(newEventsSize);
            return;
        }

        LogEvent* sourceEvent = &sourceEvents[cur].Cast<LogEvent>();
        if (sourceEvent->GetContents().empty()) {
            continue;
        }
        StringView sourceVal = sourceEvent->GetContent(mSourceKey);

        if (!isPartialLog) {
            // it is impossible to enter this state if only end pattern is given
            boost::regex regex;
            if (mMultiline.GetStartPatternReg() != nullptr) {
                regex = *mMultiline.GetStartPatternReg();
            } else {
                regex = *mMultiline.GetContinuePatternReg();
            }

            if (BoostRegexMatch(sourceVal.data(), sourceVal.size(), regex, exception)) {
                events.emplace_back(sourceEvent);
                begin = cur;
                isPartialLog = true;
            } else {
                HandleUnmatchedLogs(sourceEvents, newSize, cur, cur);
            }
        } else {
            if (mMultiline.GetContinuePatternReg() != nullptr
                && BoostRegexMatch(
                    sourceVal.data(), sourceVal.size(), *mMultiline.GetContinuePatternReg(), exception)) {
                events.emplace_back(sourceEvent);
                continue;
            }
            if (mMultiline.GetEndPatternReg() != nullptr) {
                events.emplace_back(sourceEvent);
                if (mMultiline.GetContinuePatternReg() != nullptr) {
                    // current line is not matched against the continue pattern, so the end pattern will decide if the
                    // current log is a match or not
                    if (BoostRegexMatch(
                            sourceVal.data(), sourceVal.size(), *mMultiline.GetEndPatternReg(), exception)) {
                        MergeEvents(events, true);
                        sourceEvents[newSize++] = std::move(sourceEvents[begin]);
                    } else {
                        HandleUnmatchedLogs(sourceEvents, newSize, begin, cur);
                        events.clear();
                    }
                    isPartialLog = false;
                } else {
                    if (BoostRegexMatch(
                            sourceVal.data(), sourceVal.size(), *mMultiline.GetEndPatternReg(), exception)) {
                        MergeEvents(events, true);
                        sourceEvents[newSize++] = std::move(sourceEvents[begin]);
                        if (mMultiline.GetStartPatternReg() != nullptr) {
                            isPartialLog = false;
                        } else {
                            // only end pattern is given, so start another log automatically
                            begin = cur + 1;
                        }
                    }
                    // no continue pattern given, and the current line in not matched against the end pattern, so wait
                    // for the next line
                }
            } else {
                if (mMultiline.GetContinuePatternReg() == nullptr
                    && !BoostRegexMatch(
                        sourceVal.data(), sourceVal.size(), *mMultiline.GetStartPatternReg(), exception)) {
                    events.emplace_back(sourceEvent);
                    continue;
                }
                MergeEvents(events, true);
                sourceEvents[newSize++] = std::move(sourceEvents[begin]);
                if (mMultiline.GetContinuePatternReg() != nullptr
                    && !BoostRegexMatch(
                        sourceVal.data(), sourceVal.size(), *mMultiline.GetStartPatternReg(), exception)) {
                    // when no end pattern is given, the only chance to enter unmatched state is when both start and
                    // continue pattern are given, and the current line is not matched against the start pattern
                    isPartialLog = false;
                } else {
                    begin = cur;
                    events.emplace_back(sourceEvent);
                }
            }
        }
    }
    // when in unmatched state, the unmatched log is handled one by one, so there is no need for additional handle here
    if (isPartialLog && begin < sourceEvents.size()) {
        if (mMultiline.GetEndPatternReg() == nullptr) {
            MergeEvents(events, true);
            sourceEvents[newSize++] = std::move(sourceEvents[begin]);
        } else {
            HandleUnmatchedLogs(sourceEvents, newSize, begin, sourceEvents.size() - 1);
        }
    }

    if (newSize == 0) {
        StringView firstLine = sourceEvents.front().Cast<LogEvent>().GetContent(mSourceKey);
        StringView LastLine = sourceEvents.back().Cast<LogEvent>().GetContent(mSourceKey);
        StringView filePath = logGroup.GetMetadata(EventGroupMetaKey::LOG_FILE_PATH_RESOLVED);
        LOG_WARNING(
            mContext->GetLogger(),
            ("no complete log found in current log event group", "treat according to param UnmatchedContentTreatment")(
                "first line", firstLine)("last line", lastLine)("#lines", ToString(sourceEvents.size()))(
                "filePath", filePath)("processor", sName)("config", mContext->GetConfigName()));
        mContext->GetAlarm().SendAlarm(REGEX_MATCH_ALARM,
                                       "no complete log found in current log event group: treat according to param "
                                       "UnmatchedContentTreatment\tfirst line: "
                                           + firstLine + "\tlast line: " + lastLine
                                           + "\t#lines: " + ToString(sourceEvents.size()) + "\tfilePath: " + filePath
                                           + "\tprocessor: " + sName + "\tconfig: " + mContext->GetConfigName(),
                                       mContext->GetProjectName(),
                                       mContext->GetLogstoreName(),
                                       mContext->GetRegion());
    }
    sourceEvents.resize(newSize);
}
```